### ISSUE TITLE
Passwords are not persisted correctly

### ISSUE THREAD

**Issue by lennybacon** [NuGet/Home#12344]
### NuGet Product Used

NuGet.exe

### Product Version

6.4.0.123

### Worked before?

no

### Impact

I'm unable to use this version

### Repro Steps & Context

When adding a password it is not persisted correctly.

1. Add a source
```PowerShell
& C:\agent\externals\nuget\nuget.exe sources add -Name "Sample" -Source "https://nuget.example.org/nuget/v3/index.json";
```

2. set a password
```PowerShell
& C:\agent\externals\nuget\nuget.exe sources update -Name "Sample" -UserName "DOMAIN\BuildAgent" -Password "Password`"withQuote" -StorePasswordInClearText
cat "$($env:USERPROFILE)\AppData\Roaming\NuGet\NuGet.Config"
```

**Issue 1: This results in an encrypted password. The help command does not indicate this behavior**

From what I traced in fiddler (decoded the basic auth header) the arguments after the password are appended as being part of the password.

3. Set a password with explicit undocumented order of arguments

```PowerShell
& C:\agent\externals\nuget\nuget.exe sources update -Name "Sample" -UserName "DOMAIN\BuildAgent"  -StorePasswordInClearText -Password "Password`"withQuote"
cat "$($env:USERPROFILE)\AppData\Roaming\NuGet\NuGet.Config"
```

**Issue 2: This results in the invalid password `<add key="ClearTextPassword" value="PasswordwithQuote" />` as it looses the quote, which should have been XML-encoded `<add key="ClearTextPassword" value="Passwordwith&quot;Quote" />`**

So I see basically two problems. The command line parser failing and the missing XML-encoding.

Just to make sure... The issues are related to encrypted passwords too. I just used the `-StorePasswordInClearText` to be able to show the results with `cat` without the hassle of decrypting the values in the config.

**Comment by zivkan** [comment]
@lennybacon out of curiosity, I wrote this little sample app:

```cs
Console.WriteLine("Number of arguments: " + args.Length);
for (int i = 0; i < args.Length; i++)
{
    Console.WriteLine("Arg " + i + ": " + args[i]);
}
Console.WriteLine("Original args: " + Environment.CommandLine);
```

Running your command line sample, I got this output:
```text
Number of arguments: 8
Arg 0: sources
Arg 1: update
Arg 2: -Name
Arg 3: Sample
Arg 4: -UserName
Arg 5: DOMAIN\BuildAgent
Arg 6: -Password
Arg 7: PasswordwithQuote -StorePasswordInClearText
Original args: "C:\src\test\cliParsing\bin\Debug\net48\cliParsing.exe"  sources update -Name Sample -UserName DOMAIN\BuildAgent -Password "PasswordwithQuote -StorePasswordInClearText"
```

So, to me it really looks like PowerShell isn't running the program with the arguments you expect. Especially given that the quote is stripped out of the "password with quote", in both `args` and `Environment.CommandLine`, I don't see how NuGet can write a custom parser to get the expected result.

I had a 60 second play around with different syntaxes, but I couldn't figure out how to get powershell to invoke my test program with the right string. However, going to `cmd.exe`, and changing the password to `"Password\"withQuote"`, I got the expected output.

**Comment by lennybacon** [comment]
I used your code:

```csharp
using System;

namespace ConsoleApp1
{
  internal class Program
  {
    static void Main(string[] args)
    {
      Console.WriteLine("Number of arguments: " + args.Length);
      for (int i = 0; i < args.Length; i++)
      {
        Console.WriteLine("Arg " + i + ": " + args[i]);
      }
      Console.WriteLine("Original args: " + Environment.CommandLine);
    }
  }
}
```

compile it and run the following command line from PowerShell (where the quote is escaped with a backtick (`):

```PowerShell
& .\ConsoleApp1.exe -Password "PasswordWith`"Quote"
```

The output is:

```
Number of arguments: 2
Arg 0: -Password
Arg 1: PasswordWith"Quote
Original args: "R:\ConsoleApp1\bin\Debug\ConsoleApp1.exe" -Password "PasswordWith\"Quote"
```

So where is the quote "stripped out"?

**Comment by zivkan** [comment]
I was using:

```text
dotnet run -f net48 -- -password "Password`"withQuote"
```

rather than invoking the .exe directly.  Therefore, my test actually found a bug in the .NET CLI!  I'll report it, and let the team's hotseat (rotating support person) deal with the bug you reported.

**Comment by kartheekp-ms** [comment]
> Issue 1: This results in an encrypted password. The help command does not indicate this behavior

That is weird but I think the root cause of the problem is `-StorePasswordInClearText` option is passed after `-Password` option. The following command succeeded. The only difference here is `-Password "Password\""withQuote"` option value followed by `-StorePasswordInClearText` option.

```
❯ .\nuget.exe sources add -Name "Sample2" -Source "https://nuget.example.rg/nuget/v3/index.json"  -ConfigFile .\nuget.config -UserName "DOMAIN\BuildAgent" -Password "Password\""withQuote" -StorePasswordInClearText

Package source with Name: Sample2 added successfully.
```
> Issue 2: This results in the invalid password <add key="ClearTextPassword" value="PasswordwithQuote" /> as it looses the quote, which should have been XML-encoded <add key="ClearTextPassword" value="Passwordwith&quot;Quote" />

The XML encoding worked properly if I set `-Password` value to `"Password\""withQuote"`.

```
.\nuget.exe sources add -Name "Sample2" -Source "https://nuget.example.rg/nuget/v3/index.json" -ConfigFile .\nuget.config -UserName "DOMAIN\BuildAgent" -Password "Password\""withQuote" -StorePasswordInClearText
```

``` xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <!--To inherit the global NuGet package sources remove the <clear/> line below -->
    <clear />
    <add key="Sample2" value="https://nuget.example.org/nuget/v3/index.json" />
  </packageSources>
  <packageSourceCredentials>
    <Sample2>
      <add key="Username" value="DOMAIN\BuildAgent" />
      <add key="ClearTextPassword" value="Password&quot;withQuote" />
    </Sample2>
  </packageSourceCredentials>
</configuration>
```

@lennybacon Can you use backslash (\\) as escaping character instead of backtick (`) as a work around?

**Comment by lennybacon** [comment]
@kartheekp-ms I can use `\` but PowerShell will not accept it. See https://serverfault.com/questions/47811/what-is-the-literal-escape-character-in-powershell

PowerShell uses a backtick for escaping.

**Comment by kartheekp-ms** [comment]
I tried executing the command from the PowerShell 7 (built on top of .NET 7) terminal. NuGet.exe sources add command worked as expected. Used backtick as escape character in the password.

![image](https://user-images.githubusercontent.com/52756182/210661986-1a4ac4c7-9cff-4baa-a509-bb6bc14baec4.png)

NuGet.exe is failing to parse the arguments correctly on PowerShell 5.1 but parses successfully on PowerShell 7. I assume you are experiencing this issue on CI agent. Is that correct? Is there any way you can use PowerShell 7 version to unblock. AFAIK, It doesn't appear to be a NuGet bug.

![image](https://user-images.githubusercontent.com/52756182/210664635-075fc44a-13d9-4409-96b8-49a28d70d64d.png)

**Comment by lennybacon** [comment]
@kartheekp-ms how do the sentences "NuGet.exe is failing ..." and "It doesn't appear to be a NuGet bug." fit together? Does it mean we have a Bug in PowerShell (5.1)?

@zivkan What are your thoughts?

Your assumption about the CI agent is correct. As the command is run during setup of the agent, I can of course use `cmd.exe` or `pwsh.exe` (7) to run nuget and add the source.

**Comment by zivkan** [comment]
There does seem to be some kind of issue between PowerShell 5.1 and .NET (Core) projects: https://github.com/dotnet/sdk/issues/29770

Testing with the test program I posted above, it appears that the dotnet cli (which includes .NET 5+ .exe's, as they either embed the .NET runtime, or they detect and load the runtime, even if `dotnet` isn't the first word in the command line) can't parse the `"` characters. It's not even available in `Environment.CommandLine`.

So, while nuget.exe (being .NET Framework) could theoretically have its CLI parser re-written (😭) to use `Environment.CommandLine` instead of `Main`'s `string[] args` (I hate to think how hard it will be to support both PS5.1 style quote handling, without breaking how PS7, CMD, and bash on Linux/Mac with mono), it looks like `dotnet nuget` commands are blocked without changes to the .NET runtime.

**Comment by aortiz-msft** [comment]
Closing as duplicate of https://github.com/dotnet/sdk/issues/29770
